presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaPageSource.java

```diff
@@ -0,0 +1,512 @@
+    private long memoryUsageBytes;
+    private boolean closed;
+
+    public InformationSchemaPageSource(Session session, Metadata metadata, AccessControl accessControl, InformationSchemaTableHandle tableHandle, List<ColumnHandle> columns)
+    {
+        this.session = session;
+        this.metadata = metadata;
+        this.accessControl = accessControl;
+        this.tableHandle = tableHandle;
+
+        SchemaTableName schemaTableName = tableHandle.getSchemaTableName();
+
+        tableType = getTableTypeBySchemaTableName(schemaTableName);
+
+        List<ColumnMetadata> columnMetadata = TABLES.get(schemaTableName).getColumns();
+        columnTypes = columnMetadata.stream()
+                .map(ColumnMetadata::getType)
+                .collect(toImmutableList());
+
+        rawPages = new LinkedList<>();
+
+        rawPageBuilder = new PageBuilder(columnTypes);
+
+        Map<String, Integer> columnNameToChannelMapping = IntStream.range(0, columnMetadata.size())
+                .boxed()
+                .collect(Collectors.toMap(i -> columnMetadata.get(i).getName(), Function.identity()));
+
+        outputPageChannels = columns.stream()
+                .map(columnHandle -> (InformationSchemaColumnHandle) columnHandle)
+                .map(columnHandle -> columnNameToChannelMapping.get(columnHandle.getColumnName()))
+                .collect(toImmutableList());
+
+        if (tableType == InformationSchemaTableType.TABLE_SCHEMATA
+                || tableType == InformationSchemaTableType.TABLE_ROLES
+                || tableType == InformationSchemaTableType.TABLE_APPLICABLE_ROLES
+                || tableType == InformationSchemaTableType.TABLE_ENABLED_ROLES) {
+            schemaPrefixes = Optional.empty();
+            tableNameFilterStrings = Optional.empty();
+            switch (tableType) {
+                case TABLE_SCHEMATA:
+                    buildPagesSchemata();
+                    break;
+                case TABLE_ROLES:
+                    buildPagesRoles();
+                    break;
+                case TABLE_APPLICABLE_ROLES:
+                    buildPagesApplicableRoles();
+                    break;
+                case TABLE_ENABLED_ROLES:
+                    buildPagesEnabledRoles();
+                    break;
+            }
+            flushPageBuilder();
+        }
+        else {
+            schemaPrefixes = Optional.of(getSchemaPrefixes().iterator());
+            if ((tableType == InformationSchemaTableType.TABLE_TABLES
+                    || tableType == InformationSchemaTableType.TABLE_VIEWS)
+                    && tableHandle.getConstraint().isPresent()) {
+                tableNameFilterStrings = filterString(tableHandle.getConstraint().get(), TABLE_NAME_COLUMN_HANDLE);
+            }
+            else {
+                tableNameFilterStrings = Optional.empty();
+            }
+        }
+
+        completedBytes = 0;
+        memoryUsageBytes = 0;
+        for (Page page : rawPages) {
```

Comment:
When the table is 'schemata', 'roles', 'applicable_roles' or 'enabled_roles', `rawPages` has contents. These tables are built once in the constructor. I put their building in the constructor to simplify code.

Reply:
I think we should get pages for these tables in the same place like for others. I don't see a reason why it is handled differently.

For example in case that these records were not actually needed there was no need to create them.

Also it does not match your commit message description. You described that you are loading data lazily, but here some data is loaded eagerly.